ipykernel/heartbeat.py

```diff
@@ -92,13 +92,17 @@ def _bind_socket(self):
     def run(self):
         """Run the heartbeat thread."""
         self.name = "Heartbeat"
-        self.socket = self.context.socket(zmq.ROUTER)
-        self.socket.linger = 1000
+
         try:
+            self.socket = self.context.socket(zmq.ROUTER)
```

Comment:
it may be a real dumb question, but could/should this be refactored (in another PR maybe) into a context manager:

```
with self.context.socket(zmq.ROUTER) as self.socket:
    ...
```

Reply:
Maybe, I was just trying to get rid of a failure in [test_merge_connection_file](https://github.com/ipython/ipykernel/blob/28d096d57e4750ab4a92a69e3c6544140b384729/tests/test_kernelapp.py#L50), which creates an `IPKernelApp`, initializes its heartbeat socket and thread, and closes the ZMQ context, leaving the thread working on a closed ZMQ context :cry: 
It was just working by chance before.